tasks/queue-granules/index.js

```diff
@@ -1,51 +1,109 @@
 'use strict';
 
 const get = require('lodash/get');
-const groupBy = require('lodash/groupBy');
-const chunk = require('lodash/chunk');
 const isNumber = require('lodash/isNumber');
+const memoize = require('lodash/memoize');
 const pMap = require('p-map');
 
 const cumulusMessageAdapter = require('@cumulus/cumulus-message-adapter-js');
-const { enqueueGranuleIngestMessage } = require('@cumulus/ingest/queue');
+const {
+  getWorkflowFileKey,
+  templateKey,
+} = require('@cumulus/common/workflows');
 const { constructCollectionId } = require('@cumulus/message/Collections');
 const { buildExecutionArn } = require('@cumulus/message/Executions');
+const { buildQueueMessageFromTemplate } = require('@cumulus/message/Build');
+const { getJsonS3Object } = require('@cumulus/aws-client/S3');
+const { sendSQSMessage } = require('@cumulus/aws-client/SQS');
+
 const {
+  collections: collectionsApi,
   providers: providersApi,
   granules: granulesApi,
 } = require('@cumulus/api-client');
-const CollectionConfigStore = require('@cumulus/collection-config-store');
 
-async function fetchGranuleProvider(prefix, providerId) {
+async function fetchGranuleProvider(event, providerId) {
+  if (!providerId || providerId === event.config.provider.id) {
+    return event.config.provider;
+  }
+
   const { body } = await providersApi.getProvider({
-    prefix,
+    prefix: event.config.stackName,
     providerId,
   });
 
   return JSON.parse(body);
 }
 
 /**
- * Group granules by collection and split into batches then split again on provider
+ * The purpose of this iterable is to avoid creating all the group and chunk arrays all
+ * at once to save heap space for large event inputs. This is done by using a sequence of
+ * generators that yield the chunks of granules one at a time.
  *
- * @param {Array<Object>} granules - list of input granules
- * @param {number} batchSize - size of batch of granules to queue
- * @returns {Array<Object>} list of lists of granules: each list contains granules which belong
- *                          to the same collection, and each list's max length is set by batchSize
+ * The *[Symbol.iterator] method loops over the collection looking for groups that have not
+ * already been yielded. Once the first granule of an unseen group is located, a new iterator is
+ * created that will search from that point in the collection of granules looking for other
+ * granules that to the same group.
```

Comment:
This is a great description! However, is it missing a word? 
```suggestion
 * granules that belong to the same group.
```